scalableminds/webknossos
# Improve performance of volume annotation tools

This PR adds a `chunkIntoBuckets` method to `BoundingBox` which is used to iterate over an area per bucket. This allows for efficient bucket-wise labeling operations when volume tracing. Additionally, the `Voxelterator` is stripped from its original features (voxel-wise iteration) and now only serves as a container for "painted voxels" (~probably makes sense to rename this class~ I renamed it to VoxelBuffer2D).
Also, the `VolumeLayer` now works on a Uint8Array instead of a nested array.

I benchmarked the performance, by measuring the labeling time for a single brush click (brush size: 219) in mag 1.
```
Before this PR:
AVG: 1046.999921875
STDDEV: 256.846660018522
MEDIAN: 1074.81298828125

With this PR:
AVG: 34.4859606975307
STDDEV: 41.8063191233203
MEDIAN: 8.896240234375
```



### URL of deployed dev instance (used for testing):
- https://performantbrush.webknossos.xyz

### Steps to test:
- brush in all magnifications and zoom to ensure that everything looks good in all mags
- brush in all viewports
- trace in all viewports
- double check overwrite behavior (modifier is CTRL to switch between overwrite-only active cell and overwrite everything; note that the modifier behavior is inverted when deleting; this has been like this before)

### Issues:
- fixes #4821 

------
